File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/G-Удаление_скобок/solution.py
import sys


def is_matching_pair(a: str, b: str) -> bool:
    return (a == '(' and b == ')') or (a == '[' and b == ']') or (a == '{' and b == '}')


def solution() -> None:
    """
    Time Complexity O(N**3)
    Memory Complexity O(N**3)
    """
    s: str = sys.stdin.readline().rstrip()
    n: int = len(s)
    dp: list[list[int]] = [[0] * n for _ in range(n)]
    path: list[list[str]] = [[''] * n for _ in range(n)]
    for ln in range(2, n + 1):
        for left in range(n - ln + 1):
            right: int = left + ln - 1
            if is_matching_pair(s[left], s[right]):
                if ln == 2:
                    dp[left][right] = 2
                    path[left][right] = s[left] + s[right]
                else:
                    dp[left][right] = 2 + dp[left + 1][right - 1]
                    path[left][right] = s[left] + path[left + 1][right - 1] + s[right]
            for k in range(right):
                if dp[left][right] < dp[left][k] + dp[k + 1][right]:
                    dp[left][right] = dp[left][k] + dp[k + 1][right]
                    path[left][right] = path[left][k] + path[k + 1][right]
    sys.stdout.write('{}\n'.format(path[0][n - 1]))

if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/G-Удаление_скобок/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

bool isMatchingPair(char a, char b) {
    return (a == '(' && b == ')') || (a == '[' && b == ']') || (a == '{' && b == '}');
}

void solution() {
    /*
    Time Complexity O(N**3)
    Memory Complexity O(N**3)
    */
    string s;
    cin >> s;
    int n = s.size();
    vector<vector<int>> dp(n, vector<int>(n, 0));
    vector<vector<string>> path(n, vector<string>(n, ""));

    for (int len = 2; len <= n; ++len) {
        for (int i = 0; i <= n - len; ++i) {
            int j = i + len - 1;
            if (isMatchingPair(s[i], s[j])) {
                if (len == 2) {
                    dp[i][j] = 2;
                    path[i][j] = string(1, s[i]) + string(1, s[j]);
                } else {
                    dp[i][j] = 2 + dp[i+1][j-1];
                    path[i][j] = string(1, s[i]) + path[i+1][j-1] + string(1, s[j]);
                }
            }
            for (int k = i; k < j; ++k) {
                if (dp[i][j] < dp[i][k] + dp[k+1][j]) {
                    dp[i][j] = dp[i][k] + dp[k+1][j];
                    path[i][j] = path[i][k] + path[k+1][j];
                }
            }
        }
    }
    cout << path[0][n-1] << endl;
}

int main() {
    solution();
}